locator/tablets.cc

```diff
@@ -878,10 +930,65 @@ std::unordered_set<sstring> tablet_aware_replication_strategy::recognized_tablet
     return opts;
 }
 
+namespace {
+
+// Create a simple map [data center -> index]. This index is used as position index in the small vector
+// in the tablet_to_dc_replication_factor_map.
+[[nodiscard]] absl::flat_hash_map<sstring, tablet_effective_replication_map::replication_factor_index_t> build_datacenter_map(const auto& data_centers_list) {
+    uint32_t datacenter_index = 0;
+    absl::flat_hash_map<sstring, tablet_effective_replication_map::replication_factor_index_t> datacenter_map;
+    datacenter_map.reserve(data_centers_list.size());
+    for (const sstring& datacenter : data_centers_list) {
+        datacenter_map.emplace(datacenter, datacenter_index);
+        ++datacenter_index;
+    }
+    return datacenter_map;
+}
+
+// Populate the `tablet_to_dc_replication_factor_map` with replication factors for each data center.
+// For each tablet index in the tablet map, it retrieves the current replica set using `get_endpoints_for_reading`
+// and counts replicas per data center.  Each data center's replication factor (RF) is saved under
+// its index according to `_datacenter_map`.
+future<utils::chunked_vector<tablet_effective_replication_map::replication_factor_list>> build_dc_replication_factor_map(const tablet_map& tablets,
+        const topology& topo, const absl::flat_hash_map<sstring, tablet_effective_replication_map::replication_factor_index_t>& datacenter_map) {
+    utils::chunked_vector<tablet_effective_replication_map::replication_factor_list> tablet_to_dc_replication_factor_map;
+    tablet_to_dc_replication_factor_map.reserve(tablets.tablet_count());
+    for (size_t tablet_index = 0; tablet_index < tablets.tablet_count(); ++tablet_index) {
+        const tablet_replica_set& replicas = tablets.get_replicas_for_reading(tablet_id{tablet_index});
+        tablet_to_dc_replication_factor_map.emplace_back(datacenter_map.size(), 0);
+        auto& token_rf_list = tablet_to_dc_replication_factor_map.back();
+        for (const auto& node : replicas) {
+            const locator::node* node_ptr = topo.find_node(node.host);
+            if (node_ptr == nullptr) {
+                continue;
+            }
+            const sstring& datacenter = topo.get_datacenter(node.host);
+            const auto it = datacenter_map.find(datacenter);
+            if (it == datacenter_map.end()) {
+                tablet_logger.debug("Could not find datacenter: {} for node: {} ", datacenter, node);
+                continue;
```

Comment:
Again, it's unsafe to continue.

Reply:
@tgrabiec  So, the failure occurs when the effective replication map is created before the node has been initialized. In this case, I get an empty topology, so data centers cannot be found. I'm not sure how to fix this correctly. Maybe we should skip the ERM creation until after the node has been initialized since it's unusable otherwise?

Reply:
We should probably fix the boot order to initialize topology before initializing user tables.

Reply:
The issue happens because the node, that's added to topology at the time of  `load_tablet_metadata` invocation has state 'none' (I guess because it has not finished initializing yet), thus topology's dc map is not updated from this node, in the following part: 
``` C++ 
// We keep location of left nodes because they may still appear in tablet replica sets
    // and algorithms expect to know which dc they belonged to. View replica pairing needs stable
    // replica indexes.
    // But we don't consider those nodes as members of the cluster so don't update dc registry.
    if (!node->left() && !node->is_none()) {        
        const auto& dc = node->dc_rack().dc;
        const auto& rack = node->dc_rack().rack;
        const auto& endpoint = node->endpoint();
        tlogger.trace("topology[{}]: add dc {}", fmt::ptr(this), dc);
        _dc_nodes[dc].emplace(node);
        _dc_rack_nodes[dc][rack].emplace(node);
        _dc_endpoints[dc].insert(endpoint);
        _dc_racks[dc][rack].insert(endpoint);
        _datacenters.insert(dc);
    }
```

Reply:
We shouldn't have tablet replicas on nodes which didn't finish bootstrap.

Reply:
If the node left the cluster, it won't be present in topology's _dc_nodes etc, but its dc_rack() should still be valid.

Reply:
> We shouldn't have tablet replicas on nodes which didn't finish bootstrap.

This is `topology_experimental_raft/test_mv_tablets_replace` test. 
As I see tablets load the replicas from the table `system.tablets` in `read_tablet_metadata`.

Reply:
> We shouldn't have tablet replicas on nodes which didn't finish bootstrap.

After spending some time with our initialization logic I'd say it does not look robust (and very easy to brake). Storage service contains links and references to objects that are not initialized yet. Some functions can be used at some stages of initialization, some cannot. There is no way of learning that information other than browsing the code. Particularly `topology_state_load` depends on `_group0` and `_cdc_gens`  which are initialized much later than Tablets, and in between there are other objects initialization that depend on Tablets, so we can't just move Tablets initialization to the point where topology is initialized. 
Also, the ERMs created in `init_non_system_keyspaces` just after the Tablets initialization are not consistent since topology at that time contains only one node - the local node, whereas tablets schema contains all the replicas that were persisted in `system.tablets`.  However, it seems not really used, so it's inconsistency is not big issue.